File: best-time-to-buy-sell-stock-iv.py
# TIME COMPLEXITY: O(n*k) in the worst case
# SPACE COMPLEXITY: O(k*k)

class Solution(object):
    def maxProfit(self, k, prices):
        """
        :type k: int
        :type prices: List[int]
        :rtype: int
        """
        # Edge Case
        if not len(prices):
            return 0

        # The total number of transactions can never exceed n/2
        # if k > n/2, buy low, sell high for every pair of consequetive values that satisfies this condition
        # Same as best time to buy sell ii
        max_profit = 0
        n = len(prices)
        if k >= n / 2:
            for i in range(1, n):
                max_profit += max(prices[i] - prices[i-1], 0)
            return max_profit

        # Same logic as best time to buy sell iii
        # instead of using a pair variables, use two arrays of size k to simulate successive transactions
        buy_i = [float('inf') for _ in range(k+1)]
        profit_i = [0 for _ in range(k+1)]

        for price in prices:
            for i in range(1, k+1):
                buy_i[i] = min(buy_i[i], price - profit_i[i-1])
                profit_i[i] = max(profit_i[i], price - buy_i[i])
        return profit_i[k]

File: best time to buy sell stock IV.java
//Time complexity:- O(N).
//Space complexity:- O(2n).

class Solution {
    public int maxProfit(int k, int[] prices) {
       int[] buy=new int[k+1];
       int[] profit=new int[k+1];
        Arrays.fill(buy,Integer.MAX_VALUE);
        for(int price:prices){
            for(int i=1;i<=k;i++){
                buy[i]=Math.min(buy[i],price-profit[i-1]);
                profit[i]=Math.max(profit[i],price-buy[i]); //similar to best buy and sell stock 3 but as here k transactions are there we will use array of size k+1 and same idea which we will offset profit of before transaction for next buy.
            }
        }
        return profit[k];
        
    }
}